File: src/bmi/wrf_hydro_nwm_bmi_driver.f90
program wrf_hydro_nwm_bmi_driver
  use bmi_wrf_hydro_nwm_mod, only: bmi_wrf_hydro_nwm
  implicit none
  class(bmi_wrf_hydro_nwm), allocatable :: wrf_hydro
  character(len=16), pointer :: model_name
  double precision :: end_time, current_time, time_step

  allocate(wrf_hydro)
  call err_check(wrf_hydro%get_component_name(model_name))
  print * , "--- Starting ", trim(model_name), " ---"

  call err_check(wrf_hydro%initialize("no config file right now"))
  call err_check(wrf_hydro%get_start_time(current_time))
  call err_check(wrf_hydro%get_end_time(end_time))

  do while (current_time < end_time)
     call err_check(wrf_hydro%update())
     call err_check(wrf_hydro%get_current_time(current_time))
  end do

  call err_check(wrf_hydro%finalize())

  print *, "--- Finished running test driver ---"


contains
  subroutine err_check(err)
    use bmi_wrf_hydro_nwm_mod, only: BMI_SUCCESS
    integer, intent(in) :: err
    if (err .ne. BMI_SUCCESS) then
       print *, "Stopping with error code", err
       stop "ERROR"
    end if
  end subroutine err_check
end program wrf_hydro_nwm_bmi_driver

File: src/CPL/BMI_cpl/wrf_hydro_nwm_bmi_driver.cpp
#include <iostream>
#include "bmi.hxx"
namespace wrf_hydro = bmi;

int main() {
  std::cout << "--- Start in C++ ---" << std::endl;

  int res;
  res = wrf_hydro::initialize("");

  double current_time, end_time;
  res = wrf_hydro::get_start_time(&current_time);
  res = wrf_hydro::get_end_time(&end_time);

  while (current_time < end_time) {
    res = wrf_hydro::update();
    res = wrf_hydro::get_current_time(&current_time);
    if ((int(current_time) % 10) == 0)
      std::cout << "time = " << current_time << std::endl;
  }

  res = wrf_hydro::finalize();
  std::cout << "--- Finished running test driver ---" << std::endl;
  return 0;
}
